### PR TITLE
Evaluates dependencies for bom after configure other gradle scripts

### PR DESCRIPTION
**Pull request by imasahiro**
Motivations:
- Sometimes a user wants to overwrite project's property (for example, group name) of subprojects
  but since bom.gradle is evaluated before evaluating gradle scripts under subprojects. As a result,
  it causes inconsistency between group name of published artifact and group name in bom file

Changes:
- Evaluates bom's dependencies after configuring Java projects

Results:
- Less inconsistency between bom file and published artifacts

Example:
```
// project tree
- module1
      |- build.gradle # Let's say build-1.gradle
- module2
     |- build.gradle # Let's say build-2.gradle
- build.gradle # Let's say build-root.gradle
- gradle.properties
```

```
# gradle.properties
group=a.b.c.root
```

```
// build-1.gradle
subprojects {
   group 'a.b.c.module1'
}
```

```
// build-2.gradle
subprojects {
   group 'a.b.c.module2'
}
```

```xml
<!-- bom.xml AS-IS -->
  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>a.b.c.root</groupId> <!-- We expect this group name should be `a.b.c.module1` but `a.b.c.root` -->
        <artifactId>module1</artifactId>
        <version>1.0.0-SNAPSHOT</version>
      </dependency>
      <dependency>
        <groupId>a.b.c.root</groupId> <!-- We expect this group name should be `a.b.c.module2` but `a.b.c.root` -->
        <artifactId>module2</artifactId>
        <version>1.0.0-SNAPSHOT</version>
      </dependency>
  </dependencyManagement>

<!-- bom.xml TO-BE -->
  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>a.b.c.module1</groupId>
        <artifactId>module1</artifactId>
        <version>1.0.0-SNAPSHOT</version>
      </dependency>
      <dependency>
        <groupId>a.b.c.module2</groupId>
        <artifactId>module2</artifactId>
        <version>1.0.0-SNAPSHOT</version>
      </dependency>
  </dependencyManagement>
...

### CODE REVIEW

**Review by ikhoon — APPROVED** [review]
Thanks, @imasahiro!

File: lib/bom.gradle [hunk 1]
```diff
@@ -1,32 +1,36 @@
 configure(projectsWithFlags('bom')) {
     apply plugin: 'java-platform'
     apply plugin: 'maven-publish'
+}
 
-    publishing {
-        publications {
-            bom(MavenPublication) {
-                from components.javaPlatform
+afterProjectsWithFlag('java') { Set<Project> projects ->
```
Comment (ikhoon) [inline lib/bom.gradle:6]: Note for other reviewers: It is good to review with hiding white space changes.

Comment (minwoox) [reply lib/bom.gradle:6]: That's a brilliant tip! 😄

**Review by minwoox — APPROVED** [review]
Thanks, @imasahiro!

**Comment by imasahiro** [comment]
We noticed a regression and was resolved by https://github.com/line/gradle-scripts/pull/124/commits/f474532b698858467dcb1034c06428148eb607e6 
I would like to say thanks to @noda-shin to address the issue

**Review by jrhee17 — APPROVED** [review]
👍 👍 👍

File: lib/bom.gradle [hunk 2]
```diff
@@ -12,16 +12,18 @@ configure(projectsWithFlags('bom')) {
         }
     }
 
-    dependencies {
-        constraints {
-            projectsWithFlags('java', 'publish').toList().sort { a, b ->
-                def groupComparison = "${a.group}".compareTo("${b.group}")
-                if (groupComparison != 0) {
-                    return groupComparison
+    afterProjectsWithFlag('java') { Set<Project> projects ->
```
Comment (ikhoon) [inline lib/bom.gradle:15]: I don't know why but the callbacks of `afterProjectsWithFlag()` was not called in the tasks' lifecycle.
It worked correctly when using `afterEvaluate()` and `projectsWithFlags()` separately.
https://github.com/line/armeria/pull/4377/files#diff-99c9f9c14309adeee494c305b31bb4774d339f731c426330858d51e3190a0f30R131-R132

Comment (imasahiro) [reply lib/bom.gradle:15]: Are you suggesting to change like this?

```diff
diff --git a/lib/bom.gradle b/lib/bom.gradle
index 97e34ca..bb6154c 100644
--- a/lib/bom.gradle
+++ b/lib/bom.gradle
@@ -12,9 +12,9 @@ configure(projectsWithFlags('bom')) {
         }
     }
 
-    afterProjectsWithFlag('java') { Set<Project> projects ->
-        dependencies {
-            constraints {
+    dependencies {
+        constraints {
+            afterEvaluate {
                 projectsWithFlags('java', 'publish').toList().sort { a, b ->
                     def groupComparison = "${a.group}".compareTo("${b.group}")
                     if (groupComparison != 0) {

```

Comment (ikhoon) [reply lib/bom.gradle:15]: Correct. It worked fine for my cases.

Comment (imasahiro) [reply lib/bom.gradle]: I see, sorry for late but applied your comment at https://github.com/line/gradle-scripts/pull/124/commits/1bc04d40e42d7620097507cf2af7c3dfd2ea9e10

Comment (ikhoon) [reply lib/bom.gradle]: Did it work as you expected? I wanted to know if this PR is ready to merge.

Comment (imasahiro) [reply lib/bom.gradle]: @ikhoon Yes, it worked well. Sorry for late reply

Comment (ikhoon) [reply lib/bom.gradle]: I'm glad it's going well. 😀

**Comment by ikhoon** [comment]
Thanks @imasahiro for taking care of this. 🙇‍♂️

**Pull request merged**